amaru/src/consensus/header_validation.rs

```diff
@@ -0,0 +1,179 @@
+use crate::{
+    consensus::{Point, ValidateHeaderEvent},
+    sync::PullEvent,
+};
+use gasket::framework::*;
+use miette::miette;
+use ouroboros::{ledger::LedgerState, validator::Validator};
+use ouroboros_praos::consensus::BlockValidator;
+use pallas_crypto::hash::Hash;
+use pallas_math::math::{FixedDecimal, FixedPrecision};
+use pallas_network::facades::PeerClient;
+use pallas_primitives::conway::Epoch;
+use pallas_traverse::MultiEraHeader;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use tracing::{info, trace};
+
+pub type UpstreamPort = gasket::messaging::InputPort<PullEvent>;
+pub type DownstreamPort = gasket::messaging::OutputPort<ValidateHeaderEvent>;
+
+/// Mocking this calculation specifically for preprod. We need to know the epoch for the slot so
+/// we can look up the epoch nonce.
+fn epoch_for_preprod_slot(slot: u64) -> u64 {
+    let shelley_epoch_length = 432000; // 5 days in seconds
+    let shelley_transition_epoch: u64 = 4;
+    let byron_protocol_consts_k: u64 = 2160;
+    let byron_epoch_length = 10 * byron_protocol_consts_k;
+    let byron_slots = byron_epoch_length * shelley_transition_epoch;
+    let shelley_slots = slot - byron_slots;
+    let epoch = (shelley_slots / shelley_epoch_length) + shelley_transition_epoch;
+
+    epoch
+}
+
+#[derive(Stage)]
+#[stage(name = "validate", unit = "PullEvent", worker = "Worker")]
+pub struct Stage {
+    peer_session: Arc<Mutex<PeerClient>>,
+
+    ledger: Arc<Mutex<dyn LedgerState>>,
+    epoch_to_nonce: HashMap<Epoch, Hash<32>>,
```

Comment:
@AndrewWestberg I feel like this should be part of the `LedgerState` trait to begin with. I would expect the ledger to be keeping track of the rolling nonce, and providing a way for the consensus to access the current nonce when needed. 

What do you think?

Reply:
Unrelated comment, but is `Hash<32>` a good type? Don't we want it to be more `Hash<Blake2b_256>` or `Hash<Blake2b<32>>`?

Reply:
@KtorZ I'm not sure if it belongs in Ledger, or in Chain.  In most of my projects I separate out the concept of the Chain (only deals with block headers), and Ledger (deals with the whole block, txns, etc...).

The rolling nonce and epoch nonce is something that's calculated purely from the headers. In CNCLI, I don't even bother with block fetch to do leaderlogs.

I like the idea of separating these concerns also because blockfetch() can be made more performant if we let it sync larger chunks of blocks than just one at a time.

For my tools that do both chainsync and blockfetch like JorManager, I separate these concerns. Chainsync runs ahead of blockfetch when catching up to the tip.  Blockfetch can then grab stuff in bigger chunks and it also takes it longer to process transactions.

Reply:
> @abailly: Unrelated comment, but is Hash<32> a good type? Don't we want it to be more Hash<Blake2b_256> or Hash<Blake2b<32>>?

At this point in Pallas, `Hash` exclusively refers to Blake2b hash digests. So the type is somewhat unambiguous. It will as soon as we introduce need for another hash algorithm. In which case I rather just swap `Hash` for `Blake2bDigest`.